### PR TITLE
Fix template selection resetting while changing runtimes in wizard

### PR DESCRIPTION
**Pull request by rli**
## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

Kapture 2020-12-08 at 11 15 44

## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/wizard/SamProjectWizard.kt [hunk 1]
```diff
@@ -88,6 +88,9 @@ abstract class SamProjectTemplate {
         }
     }
 
+    // defined so that we can restore the template selection when the runtime selection changes
+    override fun equals(other: Any?) = other?.let { this::class.java == it::class.java } ?: super.equals(other)
```
Comment (abrooksv) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/wizard/SamProjectWizard.kt:92]: The equals check feels odd, is it really needed?

We only create 1 SDK template ever and store them in lists, so why is the default identity check not working?

**Comment by kiiadi** [comment]
while we're at it - can we make the java template default to gradle instead of maven?

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/wizard/SamProjectWizard.kt (hunk 1 above)
Comment (rli) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/wizard/SamProjectWizard.kt:92]: when setting the selected item with `templateComboBox.selectedItem = selectedTemplate`, a equality check runs against everything in the model to figure out which one

**Comment by rli** [comment]
> while we're at it - can we make the java template default to gradle instead of maven?

done

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/java/JavaSamProjectWizard.kt [hunk 2]
```diff
@@ -32,12 +32,12 @@ class JavaSamProjectWizard : SamProjectWizard {
...
-        SamHelloWorldMaven(),
         SamHelloWorldGradle(),
-        SamEventBridgeHelloWorldMaven(),
+        SamHelloWorldMaven(),
         SamEventBridgeHelloWorldGradle(),
-        SamEventBridgeStarterAppMaven(),
-        SamEventBridgeStarterAppGradle()
+        SamEventBridgeHelloWorldMaven(),
+        SamEventBridgeStarterAppGradle(),
+        SamEventBridgeStarterAppMaven()
```
Comment (kiiadi) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/java/JavaSamProjectWizard.kt:40]: 👍

**Review by kiiadi — APPROVED** [review]

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/wizard/SamProjectWizard.kt (hunk 1 above)
Comment (rli) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/wizard/SamProjectWizard.kt:92]: to be more specific, the specific instances in the model change after a runtime change:

https://github.com/aws/aws-toolkit-jetbrains/blob/a110ba448825fb7b3bff298d3e19bebad1be807d/jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/wizard/SamInitSelectionPanel.kt#L117-L124

Comment (rli) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/wizard/SamProjectWizard.kt]: swapped to be keyed off `displayName()` instead and added hashCode() override

**Review by abrooksv — APPROVED** [review]

**Pull request merged**